### ISSUE TITLE
Remove absolute paths from config files and source code

### ISSUE THREAD

**Issue by korydraughn** [irods/irods#6283]
Having absolute paths in the code base makes it difficult to support certain installation schemes.

See https://github.com/irods/irods/pull/6275#discussion_r835961113 for more details.

**Comment by korydraughn** [comment]
@SwooshyCueb Any thoughts on this?

I tempted to say this is complete, but based on the linked comment, you may be aware of other things that need attention before we close this.

**Comment by SwooshyCueb** [comment]
There's still tons in the test suite and other scripts

**Comment by SwooshyCueb** [comment]
Additionally, as far as I am aware, the schema files still contain absolute paths, which was what this issue was originally opened to address

**Comment by trel** [comment]
most/all of the schemas now have a .in file... which are populated/converted upon installation I believe...

```
    "$id": "file:///@IRODS_CONFIGURATION_SCHEMA_PREFIX@@IRODS_HOME_DIRECTORY@/configuration_schemas/v4/server.json",
```

**Comment by korydraughn** [comment]
Correct. We ship with some defaults, but users are free to change the values during or after setup.

What would be the alternative for this?

**Comment by SwooshyCueb** [comment]
> most/all of the schemas now have a .in file... which are populated/converted upon installation I believe...

Referring back to the linked discussion, we were previously in agreement that having the absolute paths in the schema files, regardless of when they're written down, is a bad idea. In fact, I'll go even further and say that I am not convinced that they should be referring to themselves by filesystem path in the first place.

**Comment by SwooshyCueb** [comment]
Also, to clarify, these values are not set during package installation. They are set by the buildsystem. This is a problem for tarball packages, relocatable RPM packages, etc

**Comment by korydraughn** [comment]
That's reasonable.

Q. How do you imagine these absolute paths be handled?
Q. Is there an established convention for this we can reference?

If you have ideas about how to make this better, please add them here.

**Comment by korydraughn** [comment]
Are you proposing we stop shipping schema files altogether? _(I'm not opposed to shipping fewer things. It's more about making sure we cover the negative space)_

What would the default be for `schema_validation_base_uri` in server_config.json?
Is it disabled on a brand new install?
What happens if internet access isn't allowed due to company policy?

Are these questions answered via the setup script? _(because we've updated it accordingly)_

**Comment by SwooshyCueb** [comment]
> Are you proposing we stop shipping schema files altogether?

No, I'm proposing we change `$id` to a network-addressable URL, but continue to ship them in the package. Whether or not there's actually something on the other end of that URL doesn't matter as far as the schema spec is concerned.

**Comment by korydraughn** [comment]
Ah, I see now.

Many of those schema files rely on `$id` to resolve dependent/sibling schema files. Changing that to a network addressable URL could break validation, I think.

**Comment by SwooshyCueb** [comment]
I'm sure there's a way around that. I'll keep looking.

Let's reduce the scope of this issue to just addressing the absolute paths in our schemas, and open a new issue for addressing absolute paths anywhere else they might exist.

**Comment by korydraughn** [comment]
Please open a new issue for addressing locations other than the schema files.

**Comment by SwooshyCueb** [comment]
As far as I can tell, the only code that does anything with json schema is in our scripts. There's nothing relevant in any of our compiled code, aside from this little bit of `_rsZoneReport`: https://github.com/irods/irods/blob/ef676952df59d31aabf2db4ec65efea8ccbbbc48/server/api/src/rsZoneReport.cpp#L226-L228

Based on what I have learned so far, here's my current plans.

TODO:
- Change `$id` in json schemas to network-addressable URIs following the pattern `https://schemas.irods.org/[category]/v[version]/[filename]`.
    - Example: `https://schemas.irods.org/configuration/v4/configuration_directory.json`
- Deploy schemas in `[irods "home" dir]/schemas/[category]/v[version]/[filename]`.[^1]
    - Example: `/var/lib/irods/schemas/configuration/v4/configuration_directory.json`
    - Alternatively, `[irods "home" dir]/schemas/[type]/[category]/v[version]/[filename]`.[^1]
        - Example: `/var/lib/irods/schemas/json/configuration/v4/configuration_directory.json`
- Remove all references to `schema_validation_base_uri`.[^2]
- Change `validate_json.py` and `json_validation.py` to not conflate schema location and schema id.
- Add tests.

Maybe TODO:
- Tell `jsonschema` how to find our schemas, likely via a registry and/or resolver.
- Change all `$ref`s in json schemas to full schema IDs.[^2]
- Add an array to `server_config.json` schema containing directories to recursively search for json schemas.[^2]
- Re-work/remove `schema_name` keys.[^2][^3]
- Re-work/remove `schema_version` keys.[^2][^3]

[^1]: At present, we shove everything and the kitchen sink in the irods "home" directory, usually `/var/lib/irods`. As we move towards Normal and Boring, we will start to move things out of `/var/lib` and into more sensible locations. The schema files will probably end up somewhere like `/usr/share/irods/schemas` or `/usr/lib/irods/schemas`.
[^2]: This will require changing json schema. Does this warrant a version bump?
[^3]: My understanding is that if a json object must reference its own schema, the convention is to put the schema id in a `$schema` key.

**Comment by trel** [comment]
we were going to turn off schemas.irods.org...

**Comment by trel** [comment]
> Does this warrant a version bump?

somewhere between 'quite likely' to 'definitely'

**Comment by SwooshyCueb** [comment]
> we were going to turn off schemas.irods.org...

Like I said, we don't have to actually host anything on those URLs unless we just want to. They're an just an identifier, like XML namespaces.

**Comment by trel** [comment]
> we don't have to actually host anything on those URLs unless we just want to

interesting.   okay, so, demo it all behaving and i'm in.

**Comment by korydraughn** [comment]
PR #7988 addresses the config schema files and covers the primary TODO items at https://github.com/irods/irods/issues/6283#issuecomment-1632967237.

The "Maybe TODOs" from https://github.com/irods/irods/issues/6283#issuecomment-1632967237 should have a dedicated issue.

Based on https://github.com/irods/irods/issues/6283#issuecomment-1624315151, the tests should have a dedicated issue too.

Splitting the work up will allow us to gradually handle things.

Thoughts?

**Comment by trel** [comment]
yes, if specific remaining things don't have to be handled together, please file them separately.

**Comment by trel** [comment]
Agreed - 4 should see little movement for schemas.   The action is in 5.

So what is the remaining work for this particular issue?

**Comment by korydraughn** [comment]
The primary work is complete. That being, removing the absolute paths from the schema files.

We have issues for iRODS 5 which cover changing where the schema files are stored and other things. See #7427.

Closing.
